### PR TITLE
Hatchbuild

### PR DESCRIPTION
**Pull request by Nagidal**
Proposal to fix https://github.com/python-semver/python-semver/issues/373 (Use hatchling build backend)

I have moved everything except for flake8 settings and pycodestyle settings from setup.cfg to pyproject.toml

I have prepared these scripts:

- `style:fmt` runs Black and isort (all files now have imports sorted by isort)
- `style:lint` runs flake8 and pycodestyle
- `docs:build` runs pdoc3 to build API documentation in the api-docs directory (added to .gitignore
- `docs:serve` runs pdoc3 local server on `localhost:8080` to observe API docs updated live as you update docstrings. However, this somehow does not work in this project and I don't know exactly how to solve the error. Maybe @tomschr can figure it out, the API documentation is rather nice
- `test:cov` runs pytest-cov
- `test:no-cov` runs pytest with --no-cov
- `test:tox_test` runs tox

Scripts can be run with `hatch run`, e.g. `hatch.run style:fmt`. 

All the test scripts fail because of `python-semver\tests\coerce.py` not being a proper Python file (invalid syntax). I don't know what the purpose of this file is, so I did not touch it. It makes tests fail even if you run then natively with tox, so repairing this is beyond the scope of this PR, methinks.

I changed the classifier `Environment :: Web Environment` to `Environment :: Console` as I think it's more fitting.

### CODE REVIEW

File: changelog.d/374.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Correct Towncrier's config entries in the :file:`pyproject.toml` file.
```
Comment (tomschr) [inline changelog.d/374.bugfix.rst:1]: This file was integrated as part of PR #375 and should _not_ appear in this branch. You should remove this file and rebase your branch. Possibily you need to resolve conflicts.

File: pyproject.toml [hunk 2]
```diff
@@ -40,42 +190,23 @@ template = "changelog.d/_template.rst"
 # issue_format = "`#{issue} <https://github.com/python-attrs/attrs/issues/{issue}>`_"
 # issue_format = ":gh:`{issue}`"
 
-  # [[tool.towncrier.type]]
```
Comment (tomschr) [inline pyproject.toml:43]: The parts are already done by PR #375. The Towncrier parts below will possibly create a conflict when you rebase.

File: src/semver/__init__.py [hunk 3]
```diff
@@ -4,36 +4,12 @@
 A Python module for semantic versioning. Simplifies comparing versions.
 """
 
-from ._deprecated import (
```
Comment (tomschr) [inline src/semver/__init__.py:7]: Why is the source code changed?

I'm not entirely sure why the change of `pyproject.toml` triggers a reformatting. It shouldn't. The source code is perfectly formatted and the black config wasn't changed. This PR is all about switching from setuptools to hatch(ling) and we should keep it as such.

Could you please check and revert these changes in the source code?

**Review by tomschr — CHANGES_REQUESTED** [review]
Thanks @Nagidal for your work, very much appreciated!

I've looked through the changes and left some comments below. Unfortunately, I saw this PR a bit too late, so I merged PR #375 already. I'm sorry! That might create conflicts in your branch. You need to rebase and resolve the conflicts. If you need help, let me know or I can do it for you.

Regarding the `tests/coerce.py`: That should be a link to `docs/advanced/coerce.py`. It is used in some doc tests. The link is currently broken, I see it has some strange characters at the end of the filename. If you correct these the tests should work again.

The overall change looks pretty good, I have only these two main concerns:

* With the current `pyproject.toml` you can't build the user documentation. We need that. :slightly_smiling_face: That issue should be fixed and I've addressed it below.

* The source codes have been changed. I would have expected that such a change wouldn't touch the source code at all. I think, this should also be fixed. This PR should NOT contain any changes to the Python code.

I would also like to see in this PR is the (user) documentation. The file `BUILDING.rst` should also be adapted.

Further ideas I would like to know, but we can do that in a later PR:

* Change the Towncrier changelog. We need another entry in `pyproject.toml` where we can build, check, and create the changelog file.
* Upload of wheel and source dist to PyPI.
* Adapt the GitHub Actions to use `hatch` instead of `tox` could be debated.

File: pyproject.toml [hunk 4]
```diff
@@ -1,34 +1,184 @@
...
+[[tool.hatch.envs.test.matrix]]
+python = ["310", "311"]
+
+[tool.coverage.run]
+branch = true
+parallel = true
+omit = [
+    # add files to exclude them from the coverage report, e.g.
+    # "src/semver/__about__.py"
+]
```
Comment (tomschr) [inline pyproject.toml:130]: If I'm not mistaken, you need to tell coverage/pytest-cov where to find the source by using `source`. Furthermore, I would add `semver/_types.py` to the list of files to be excluded from coverage (it contains just type definitions):


```suggestion
[tool.coverage.run]
source = [ "src/semver/*", ]
branch = true
parallel = true
omit = [
    # add files to exclude them from the coverage report, e.g.
    "src/semver/_types.py",
]
```

File: pyproject.toml [hunk 5]
```diff
@@ -1,34 +1,184 @@
...
+[tool.hatch.envs.test]
+dependencies = [
+    "pytest-cov",
+    "tox",
+]
+
+[tool.hatch.envs.test.scripts]
+cov = "pytest -vx"
+no-cov = "cov --no-cov"
+tox_test = "tox"
```
Comment (tomschr) [inline pyproject.toml:119]: I suspect something is missing here. When I run tox (or pytest), I can run it either tests against all defined Python version or a specific version. Running `tox` alone would always run _all_ Python versions. That's a time consuming task and you maybe want to restrict it to run against Python 3.7 only.

**Review by tomschr — CHANGES_REQUESTED** [review]
I found some other minor issues and left some questions.

File: pyproject.toml [hunk 6]
```diff
@@ -1,34 +1,184 @@
...
+name = "semver"
+description = "Python helper for Semantic Versioning (https://semver.org)"
+readme = "README.rst"
+requires-python = ">=3.7"
+authors = [
+    { name = "Kostiantyn Rybnikov", email = "k-bx@k-bx.com" },
+]
+maintainers = [
+    { name = "Sebastien Celles", email = "s.celles@gmail.com" },
+    { name = "Tom Schraitle" },
```
[the remark this answers is not in the document]
Comment (Nagidal) [reply pyproject.toml:17]: That's not necessary. According to PEP 621:

> These fields accept an array of tables with 2 keys: name and email. [...] Both keys are optional.

File: changelog.d/373.trivial.rst [hunk 7]
```diff
@@ -0,0 +1 @@
+Changed build backend to hatchling
```
[the remark this answers is not in the document]
Comment (Nagidal) [reply changelog.d/373.trivial.rst:1]: Ok, so I guess I should leave the classifier change out of the scope of this PR. We would probably make a separate issue for if and update the classifiers in a dedicated PR.

File: pyproject.toml [hunk 8]
```diff
@@ -1,34 +1,184 @@
...
-build-backend = "setuptools.build_meta"
 
+[tool.hatch.envs.docs]
+dependencies = [
+    "pdoc3"
+]
+
+[tool.hatch.envs.docs.scripts]
+build = "pdoc --html --output-dir api-docs src/semver --force"
+serve = "pdoc --http : python-semver"
```
[the remark this answers is not in the document]
Comment (Nagidal) [reply pyproject.toml:108]: Sorry, I forgot and overlooked that python-semver's documentation is made by sphinx. I will get rid of pdoc3 and bring in the sphinx API doc.

File: pyproject.toml [hunk 9]
```diff
@@ -1,34 +1,184 @@
...
+    # add files to exclude them from the coverage report, e.g.
+    # "src/semver/__about__.py"
+]
+
+[tool.coverage.report]
+exclude_lines = [
+    "no cov",
+    "if __name__ == .__main__.:",
+    "if TYPE_CHECKING:",
+]
```
[the remark this answers is not in the document]
Comment (Nagidal) [reply pyproject.toml:137]: Here I did not check what python-semver's code actually uses to mark excluded lines. The "no cov" setting was copy-pasted from one of my pyproject.tomls. We shall of course have here the same comments which python-semver uses.

File: changelog.d/374.bugfix.rst (hunk 1 above)
Comment (Nagidal) [reply changelog.d/374.bugfix.rst:1]: I agree. Will do.

File: src/semver/__init__.py (hunk 3 above)
Comment (Nagidal) [reply src/semver/__init__.py:7]: I accidentally ran the `style:fmt` script, which runs [isort](https://pycqa.github.io/isort/), and it reformatted (and sorted) the import statements. Using isort would be another suggestion, but again, out of scope of this PR. I'll revert these.

File: pyproject.toml [hunk 10]
```diff
@@ -1,34 +1,184 @@
...
+    "tox",
+]
+
+[tool.hatch.envs.test.scripts]
+cov = "pytest -vx"
+no-cov = "cov --no-cov"
+tox_test = "tox"
+
+[[tool.hatch.envs.test.matrix]]
+python = ["310", "311"]
```
[the remark this answers is not in the document]
Comment (Nagidal) [reply pyproject.toml:122]: I thought I would sketch up just a little test matrix in `pyproject.toml` for the time being, to see how it works for you and your Github workflows. All tests and their environments neatly set up in `tox` and it is not necessary to migrate to a hatch test matrix right away. However, we can do this. I shall update the PR with a [3.7, ... 3.11] matrix.

File: pyproject.toml [hunk 11]
```diff
@@ -1,34 +1,184 @@
...
+exclude_lines = [
+    "no cov",
+    "if __name__ == .__main__.:",
+    "if TYPE_CHECKING:",
+]
 
+[tool.mypy]
+# the mypy settings go here
+# To have the `py.typed` file installed with the package we had to include it
+# in the build metadata (see [tool.hatch.build])
```
[the remark this answers is not in the document]
Comment (Nagidal) [reply pyproject.toml:142]: I've never used mypy, and I did not look thoroughly enough where your current mypy options were until now. But of course, we shall have these here now.

File: pyproject.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (tomschr) [reply pyproject.toml:108]: All good, absolutely no problem. :people_hugging: That why we have a review. :slightly_smiling_face:

File: pyproject.toml (hunk 9 above)
[the remark this answers is not in the document]
Comment (tomschr) [reply pyproject.toml:137]: Hehe, no problem. I assumed it was copied from somewhere. You couldn't know these details, so no worries.

File: pyproject.toml [hunk 12]
```diff
@@ -1,34 +1,184 @@
...
+
+[tool.pytest.ini_options]
+norecursedirs = ".git build .env/ env/ .pyenv/ .tmp/ .eggs/ venv/"
```
[the remark this answers is not in the document]
Comment (Nagidal) [reply pyproject.toml:145]: I couldn't tell from [pytest's documentation](https://docs.pytest.org/en/7.1.x/reference/reference.html#confval-norecursedirs) whether `norecursedirs` accepts a list, or if it all must be one string (the same is not sure for [addopts](https://docs.pytest.org/en/7.1.x/reference/reference.html#confval-addopts)). I shall test if these config options understand lists and see.

File: src/semver/__init__.py (hunk 3 above)
Comment (tomschr) [reply src/semver/__init__.py:7]: Again, no problem.
Actually, I was thinking about isort too. But as you said, we can do that in another PR. :+1:

File: pyproject.toml (hunk 4 above)
Comment (Nagidal) [reply pyproject.toml:130]: ok, will do.

File: pyproject.toml (hunk 5 above)
Comment (Nagidal) [reply pyproject.toml:119]: umm, sure. I think it would be convenient to use the `test` environment for testing only with Python 3.7, and have another test environment, e.g. `testall` with the complete matrix of Python versions.

For running tox with all its regular tests I will prepare a `hatch run tox:test` script.

File: pyproject.toml (hunk 2 above)
Comment (Nagidal) [reply pyproject.toml:43]: yup, I shall exclude the #375 changes from this PR. Until now I was only doing PRs for my one-man show projects, never interfering with anyone else's changes upstream, So I picked up these nasty habits of mixing in other unrelated changes into the PR. I'll correct this. Happy to work with you on a more complex project and learn to produce some more disciplined PRs.

**Comment by Nagidal** [comment]
I will try to correct this tomorrow, or by Monday latest. Thanks for all the suggestions and push-backs.

**Comment by tomschr** [comment]
Thank you very much. You're welcome. Take your time, no need to hurry. :slightly_smiling_face:
